File: Python/AddNumbersWithAccumulator.py
from pyspark import SparkContext, SparkConf, Accumulator


def main() -> None:
    """
    Python program that uses Apache Spark to sum a list of numbers
    """
    spark_conf = SparkConf().setAppName("AddNumbers").setMaster("local[4]")
    spark_context = SparkContext(conf=spark_conf)

    logger = spark_context._jvm.org.apache.log4j
    logger.LogManager.getLogger("org").setLevel(logger.Level.WARN)

    data = [1, 2, 3, 4, 5, 6, 7, 8]
    distributed_data = spark_context.parallelize(data)

    accumulator = spark_context.accumulator(0)

    def sum_numbers(n1, n2):
        accumulator.add(1)
        return n1 + n2


    total = distributed_data.reduce(lambda s1, s2: sum_numbers(s1, s2))

    print("The sum is " + str(total))
    print("The number of reduce operations is: " + str(accumulator.value))


if __name__ == '__main__':
    main()

File: Java/AddNumbersWithAccumulator.java
package org.masterinformatica.spark;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.Accumulator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.util.LongAccumulator;

import java.util.Arrays;
import java.util.List;

/**
 * Program that sums a list of numbers using Apache Spark
 *
 * @author Antonio J. Nebro
 */
public class AddNumbersWithAccumulator {
    public static void main(String[] args) {
        Logger.getLogger("org").setLevel(Level.OFF);

        SparkConf sparkConf = new SparkConf()
                .setAppName("Accumulators")
                .setMaster("local[8]");
        JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);

        List<Integer> integerList = Arrays.asList(new Integer[]{1, 2, 3, 4, 5, 6, 7, 8});

        LongAccumulator accum = sparkContext.sc().longAccumulator();

        int sum = sparkContext
                .parallelize(integerList)
                .reduce((integer, integer2) -> {
                    accum.add(1);
                    return integer + integer2;
                });

        System.out.println("The sum is: " + sum);
        System.out.println("Number of reduces: " + accum.value());

        // Step 7: stop the spark context
        sparkContext.stop();
    }
}
